### PR TITLE
[autoscaler] Experimental support for local / on-prem clusters

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This adds some experimental (undocumented) support for launching Ray on existing nodes. You have to provide the head ip, and the list of worker ips.

There are also a couple additional utils added for rsyncing files and port-forward.

### CODE REVIEW

**Comment by ericl** [comment]
cc @hartikainen

File: python/ray/scripts/scripts.py [hunk 1]
```diff
@@ -445,6 +445,32 @@ def attach(cluster_config_file, start, cluster_name):
...
+@click.command()
+@click.argument("cluster_config_file", required=True, type=str)
+@click.argument("source", required=True, type=str)
+@click.argument("target", required=True, type=str)
+@click.option(
+    "--cluster-name",
+    required=False,
+    type=str,
+    help=("Override the configured cluster name."))
+def rsync_down(cluster_config_file, source, target, cluster_name):
```
Comment (richardliaw) [inline python/ray/scripts/scripts.py]: Can you add some descriptors for these functions?

File: doc/source/autoscaling.rst [hunk 2]
```diff
@@ -89,11 +89,21 @@ You can use ``ray attach`` to attach to an interactive console on the cluster.
 Port-forwarding applications
 ----------------------------
 
-To run connect to applications running on the cluster (e.g. Jupyter notebook) using a web browser, you can forward the port to your local machine using SSH:
+To run connect to applications running on the cluster (e.g. Jupyter notebook) using a web browser, you can use the port-forward option for ``ray exec``:
```
Comment (richardliaw) [inline doc/source/autoscaling.rst]: Can you leave a note that the port opened on the local machine is the same as the port forwarded on the remote machine?

File: python/ray/autoscaler/local/example-full.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    ssh_private_key: ~/.ssh/id_rsa
+head_node: {}
+worker_nodes: {}
+file_mounts:
+     "/tmp/ray_sha": "/YOUR/LOCAL/RAY/REPO/.git/refs/heads/YOUR_BRANCH"
+setup_commands: []
+head_setup_commands: []
+worker_setup_commands: []
+setup_commands:
+    - source activate ray && test -e ray || git clone https://github.com/YOUR_GITHUB/ray.git
```
Comment (richardliaw) [inline python/ray/autoscaler/local/example-full.yaml]: why not just a pip install?

File: python/ray/autoscaler/local/node_provider.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+from filelock import FileLock
+import json
+import os
+import socket
+
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import TAG_RAY_NODE_TYPE
+
+
+class ClusterState(object):
```
Comment (richardliaw) [inline python/ray/autoscaler/local/node_provider.py]: Can you provide some docstrings for this class?

**Review by richardliaw — APPROVED** [review]
Overall looks good, would be nice to have some documentation. 

Also are there going to be tests for this?

**Comment by robertnishihara** [comment]
cc @devin-petersohn @pschafhalter

File: python/ray/autoscaler/local/example-full.yaml (hunk 3 above)
Comment (robertnishihara) [reply python/ray/autoscaler/local/example-full.yaml]: These setup commands don't seem to actually install ray, do they?

File: python/ray/autoscaler/local/node_provider.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    def __init__(self, lock_path, save_path, provider_config):
+        self.file_lock = FileLock(lock_path)
+        self.save_path = save_path
+
+        with self.file_lock:
+            if os.path.exists(self.save_path):
+                workers = json.loads(open(self.save_path).read())
+            else:
+                workers = {}
+            print("Loaded cluster state", workers)
```
Comment (robertnishihara) [inline python/ray/autoscaler/local/node_provider.py]: Should we be using the logging module for new print statements?

File: doc/source/autoscaling.rst [hunk 6]
```diff
@@ -89,11 +89,21 @@ You can use ``ray attach`` to attach to an interactive console on the cluster.
...
 
 .. code-block:: bash
 
-    $ ssh -L 8899:localhost:8899 -i <key> <user>@<addr> 'source ~/anaconda3/bin/activate tensorflow_p36 && jupyter notebook --port=8899'
+    $ ray exec cluster.yaml --port-forward=8899 'source ~/anaconda3/bin/activate tensorflow_p36 && jupyter notebook --port=8899'
+
+Manually synchronizing files
+----------------------------
+
+To download or upload files to the cluster head node, use ``ray rsync_down`` or ``ray rsync_up``:
```
Comment (robertnishihara) [inline doc/source/autoscaling.rst]: Is syncing to/from the head node the primary use case, as opposed to syncing to/from all nodes?

**Comment by robertnishihara** [comment]
@ericl does this require any special instructions for what IP addresses to use when you have a public/private IP address distinction (like on EC2)?

File: python/ray/autoscaler/local/node_provider.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/autoscaler/local/node_provider.py]: we can do that separately in a PR (#2628)

File: doc/source/autoscaling.rst (hunk 6 above)
Comment (ericl) [reply doc/source/autoscaling.rst]: Yeah, handling other nodes is out of scope here.

File: python/ray/autoscaler/local/example-full.yaml (hunk 3 above)
Comment (ericl) [reply python/ray/autoscaler/local/example-full.yaml]: This example is for dev only for now...

File: doc/source/autoscaling.rst (hunk 2 above)
Comment (ericl) [reply doc/source/autoscaling.rst]: Done.

File: python/ray/scripts/scripts.py (hunk 1 above)
Comment (ericl) [reply python/ray/scripts/scripts.py]: That's not currently done in this file. Do those show up in ray --help?

**Comment by ericl** [comment]
@robertnishihara I'm assuming all nodes have just one IP for now.

**Comment by ericl** [comment]
Merging so we can start testing.

**Pull request merged**